### PR TITLE
[GSoC][TMVA][SOFIE] Serialisation of RModel

### PR DESCRIPTION
**Pull request by sanjibansg**
### Description
This PR implements serialisation of RModel by adding a custom streamer in class RModel and modifying the definition of structure TMVA::Experimental::SOFIE::InitializedTensor.

### Checklist:
- [x] tested changes locally
  
### Tests  
- [x] Tests for Serialisation of Custom Models  

### Example usage 
**Writing the RModel object** 
```
using namespace TMVA::Experimental;
TFile file("model.root","CREATE");
SOFIE::RModelParser_ONNX parser;
SOFIE::RModel model = parser.Parse("Linear_2.onnx");
model.Write("model");
file.Close();
```

**Reading the RModel object**
```
using namespace TMVA::Experimental;
TFile file("model.root","READ");
SOFIE::RModel *model;
file.GetObject("model",model);
file.Close();
model->Generate();
model->OutputGenerated("model.hxx");
```

### CODE REVIEW

**Comment by jalopezg-r00t** [comment]
@phsft-bot build

**Comment by bellenot** [comment]
@sanjibansg please fork also [roottest](https://github.com/root-project/roottest) to prevent timeouts on the Windows nodes

**Comment by lmoneta** [comment]
@sanjibansg : can you please rebase it to fix the conflict file , happening after merging your other PR on restructuring Sofie. Thank you

**Comment by sanjibansg** [comment]
> @sanjibansg : can you please rebase it to fix the conflict file , happening after merging your other PR on restructuring Sofie. Thank you

@lmoneta Resolved the merge conflict.

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu16/nortcxxmod with flags -Dtmva-sofie=On

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2004/default with flags -Dtmva-sofie=On

File: tmva/sofie/inc/TMVA/SOFIE_common.hxx [hunk 1]
```diff
@@ -50,8 +49,34 @@ std::size_t ConvertShapeToLength(std::vector<size_t> shape);
 struct InitializedTensor{
    ETensorType type;
    std::vector<std::size_t> shape;
-   std::shared_ptr<void> data;
-   //void* data;
+   std::shared_ptr<void> data;     //! Transient
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/SOFIE_common.hxx:52]: small comments: 
we should maybe use ROOT naming conventions:
- starting with `f` the data members
- capitalise member functions
  `castSharedToPersistent`  -> `CastSharedToPersistent`

File: tmva/sofie/test/CMakeLists.txt [hunk 2]
```diff
@@ -19,46 +19,93 @@ endif()
...
+	${SOFIE_PARSERS_DIR}/src/RModelParser_ONNX.cxx
+	${PROTO_SRCS}
+)
+target_include_directories(emitFromROOT PRIVATE
+  ../inc
+  ${SOFIE_PARSERS_DIR}/inc
+  ../../tmva/inc
+  ${CMAKE_CURRENT_BINARY_DIR}   # this is for the protobuf headerfile
+  ../../../core/foundation/inc
+  ${CMAKE_BINARY_DIR}/include   # this is for RConfigure.h
```
Comment (lmoneta) [inline tmva/sofie/test/CMakeLists.txt]: use ginclude instead of include

**Review by lmoneta — APPROVED** [review]
Looks good, just a small comment on the names of member function and data members

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu18.04/default with flags -Dtmva-sofie=On

File: tmva/sofie/test/CMakeLists.txt (hunk 2 above)
Comment (sanjibansg) [reply tmva/sofie/test/CMakeLists.txt]: Changed include to ginclude

File: tmva/sofie/inc/TMVA/SOFIE_common.hxx (hunk 1 above)
Comment (sanjibansg) [reply tmva/sofie/inc/TMVA/SOFIE_common.hxx]: Made the changes in names of data members & member functions of struct InitializedTensor

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2004/default with flags -Dtmva-sofie=On

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu18.04/default with flags -Dtmva-sofie=On

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu18.04/default, ROOT-ubuntu2004/default with flags -Dtmva-sofie=On

**Review by lmoneta — APPROVED** [review]
The Warnings are caused by using an old protobuf version in the build nodes. When using protobuf version 3.17 they should be fixed. 
The PR can then be merged.
Thank you @sanjibansg for this contribution!

**Pull request merged**